[Host.Outbox] Clean up in batches

EtherZa opened the pull request:
Changes:
* Extracts `OutboxCleanUpTask` from `OutboxSendingTask`
* Clean up routing occurs outside of send loop and on SMB start up
* Clean up occurs in recurring batches until `Outbox` table has been cleaned
* Clean up occurs with a low `deadlock_priority` so as to kill service instead of publishing queries

zarusz commented:
Hi @EtherZa,

This is a valuable enhancement—great work!

Regarding the outbox integration tests, I wanted to share some context: our integration tests utilize a shared Azure SQL instance, meaning all test runs access the same database. Previously, when tests relied on the SQL Server container image provided by the build agent, we encountered significant unresponsiveness, leading to flaky outbox test results. I hope this insight is helpful. Please let me know if there's anything I can assist you with.​

EtherZa commented:
Hi @zarusz. That was quite the adventure but should hopefully be it. Please let me know your thoughts when you have a chance.

EtherZa commented:
FYI: I'm reviewing the indexes. I think it can be better.

EtherZa commented:
Thanks @zarusz. I'm done. Please review when you are able to.

zarusz commented on src/SlimMessageBus.Host.Outbox/Services/OutboxCleanUpTask.cs:
```diff
@@ -0,0 +1,185 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
```
These should come from implicit usings, so please remove them.

EtherZa replied:
Done.

zarusz commented on src/SlimMessageBus.Host.Outbox/Services/OutboxCleanUpTask.cs:
```diff
@@ -0,0 +1,185 @@
+
+using Microsoft.Extensions.Hosting;
+
+using SlimMessageBus;
```
This using seems redundant

EtherZa replied:
Done.

zarusz commented on src/SlimMessageBus.Host.Outbox/Services/OutboxCleanUpTask.cs:
```diff
@@ -0,0 +1,185 @@
+    private readonly TimeProvider _timeProvider;
+    private readonly IHostApplicationLifetime _hostApplicationLifetime;
+    private readonly IServiceProvider _serviceProvider;
+    private readonly SemaphoreSlim _slim;
```
how about _semaphore?

zarusz commented on src/SlimMessageBus.Host.Outbox/Services/OutboxCleanUpTask.cs:
```diff
@@ -0,0 +1,185 @@
+            Debug.Assert(_cts == null);
+
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(_hostApplicationLifetime.ApplicationStopping);
+            _executingTask = CleanUpLoop(_cts.Token);
```
Wouldn't it be more optimal to use the [System.Threading.Timer](https://learn.microsoft.com/en-us/dotnet/standard/threading/timers) primitive to ask it to run with a specific interval, and do cleanup at those moments, and avoid a `Task.Delay()`?

EtherZa replied:
No, they pretty much do the same thing. [`Task.Delay`](https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5809) (at least when using [`TimeProvider`](https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/libraries/Common/src/System/TimeProvider.cs#L11)) uses a [`TimerQueueTimer`](https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs#L451) under the hood.

![image](https://github.com/user-attachments/assets/ed14651f-43d0-4aa6-b8bb-e7559b6369dc)

which is the same implementation that [System.Threading.Timer](https://github.com/dotnet/runtime/blob/d17420560540e4184a78f431de8341f747068c4f/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs#L908) uses

![image](https://github.com/user-attachments/assets/84da0511-98ac-4838-b77d-856e187dac75)

zarusz replied:
Ok then

zarusz commented on src/Tests/SlimMessageBus.Host.Outbox.Test/SlimMessageBus.Host.Outbox.Test.csproj:
```diff
@@ -2,7 +2,11 @@
 	<Import Project="../Host.Test.Properties.xml" />
 
-	<ItemGroup>
+  <ItemGroup>
```
Ident, reformat

EtherZa replied:
Done.

zarusz commented:
Nice work, @EtherZa! I’ve added a suggestion and a few minor comments - let me know what you think.

EtherZa commented:
> Nice work, @EtherZa! I’ve added a suggestion and a few minor comments - let me know what you think.

Thanks. I have made the changes you requested and added some detail arounnd `Task.Delay` with `TimeProvider`. tldr; it's the same as `System.Threading.Timer` :)

EtherZa commented:
Hi @zarusz. Do you have any idea when this may escape to nuget.org?

EtherZa commented:
That sounds like a superb reason! Have a fantastic time
